### PR TITLE
#49: Show inconsistencies between data and metadata

### LINKED ISSUE #49: Show inconsistencies between data and metadata
## Description
Show inconsistencies between data and metadata

### PR DESCRIPTION
**Pull request by jozefbakus**
Implements: #49 Show inconsistencies between data and metadata

### CODE REVIEW

File: src/main/scala/za/co/absa/spark_metadata_tool/model/MetadataRecord.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+package za.co.absa.spark_metadata_tool.model
+
+import org.apache.hadoop.fs.Path
+
+case class MetadataRecord(
+  path: Path,
+  action: String
+)
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/spark_metadata_tool/model/MetadataRecord.scala]: nit: no newline at the end of the file

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala [hunk 2]
```diff
@@ -101,6 +106,48 @@ object Application extends App {
...
+        notDeletedData.isEmpty && missingData.isEmpty && unknownData.isEmpty
+
+      if(isDataIssueDetected) {
+        logger.info("No issue detected in data and metadata")
+      } else {
+        logger.error("Data issue detected")
+      }
+
+      if(notDeletedData.nonEmpty) logger.error(s"The data that should have been deleted: $notDeletedData")
+      if(missingData.nonEmpty) logger.error(s"Missing data: $missingData")
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/spark_metadata_tool/Application.scala]: I would format the output a bit nicer, e.g. every element on a new line

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala [hunk 3]
```diff
@@ -101,6 +106,48 @@ object Application extends App {
...
+      val isDataIssueDetected =
+        notDeletedData.isEmpty && missingData.isEmpty && unknownData.isEmpty
+
+      if(isDataIssueDetected) {
+        logger.info("No issue detected in data and metadata")
+      } else {
+        logger.error("Data issue detected")
+      }
+
+      if(notDeletedData.nonEmpty) logger.error(s"The data that should have been deleted: $notDeletedData")
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/spark_metadata_tool/Application.scala]: `The data` -> `Data`

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala [hunk 4]
```diff
@@ -101,6 +106,48 @@ object Application extends App {
...
+    config: AppConfig,
+    io: FileManager,
+    tool: MetadataTool
+  ): Either[AppError, Unit] = {
+    val dataPath =  config.path
+    val metaPath = new Path(s"$dataPath/$SparkMetadataDir")
+
+    for {
+      metaPaths       <- io.listFiles(metaPath)
+      usedMetaFiles   <- tool.filterLastCompact(metaPaths)
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/spark_metadata_tool/Application.scala:119]: Getting this error when the file `hadoop-mapreduce.jobsummary.log` is present in the `_spark_metadata` dir.

```
2022-04-25 11:50:34,580 ERROR [Application] Couldn't parse name `Some(hadoop-mapreduce)` of the file hadoop-mapreduce.jobsummary.log, expected numeric value
java.lang.NumberFormatException: For input string: "hadoop-mapreduce"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Long.parseLong(Long.java:589)
	at java.lang.Long.parseLong(Long.java:631)
	at scala.collection.StringOps$.toLong$extension(StringOps.scala:923)
	at za.co.absa.spark_metadata_tool.model.MetadataFile$.$anonfun$fromPath$2(MetadataFile.scala:37)
	at za.co.absa.spark_metadata_tool.model.MetadataFile$.$anonfun$fromPath$2$adapted(MetadataFile.scala:37)
	at scala.Option.map(Option.scala:242)
	at za.co.absa.spark_metadata_tool.model.MetadataFile$.$anonfun$fromPath$1(MetadataFile.scala:37)
	at scala.util.Try$.apply(Try.scala:210)
	at za.co.absa.spark_metadata_tool.model.MetadataFile$.fromPath(MetadataFile.scala:37)
	at za.co.absa.spark_metadata_tool.MetadataTool.$anonfun$filterLastCompact$1(MetadataTool.scala:222)
	at cats.data.Chain$.$anonfun$traverseViaChain$3(Chain.scala:803)
	at cats.Eval$.loop$1(Eval.scala:320)
	at cats.Eval$.cats$Eval$$evaluate(Eval.scala:363)
	at cats.Eval$FlatMap.value(Eval.scala:284)
	at cats.data.Chain$.traverseViaChain(Chain.scala:825)
	at cats.instances.SeqInstances$$anon$1.traverse(seq.scala:95)
	at cats.instances.SeqInstances$$anon$1.traverse(seq.scala:14)
	at cats.Traverse$Ops.traverse(Traverse.scala:181)
	at cats.Traverse$Ops.traverse$(Traverse.scala:180)
	at cats.Traverse$ToTraverseOps$$anon$3.traverse(Traverse.scala:206)
	at za.co.absa.spark_metadata_tool.MetadataTool.filterLastCompact(MetadataTool.scala:222)
	at za.co.absa.spark_metadata_tool.Application$.$anonfun$compareMetadataWithData$1(Application.scala:119)
	at scala.util.Either.flatMap(Either.scala:352)
	at za.co.absa.spark_metadata_tool.Application$.compareMetadataWithData(Application.scala:118)
```

Not sure if that's intended or not.

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala [hunk 5]
```diff
@@ -101,6 +106,48 @@ object Application extends App {
...
+      metaRecords     <- usedMetaFiles.flatTraverse(metaFile => tool.getMetaRecords(metaFile.path))
+      filesInInputDir <- tool.listFilesRecursively(dataPath)
+      dataFiles       <- filesInInputDir.filter(_.toString.endsWith(".parquet")).asRight
+    } yield {
+      val ADD = "add"
+      val DELETE = "delete"
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/spark_metadata_tool/Application.scala]: Why uppercase?

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala [hunk 6]
```diff
@@ -101,6 +106,48 @@ object Application extends App {
...
+      val metaRecordsByAction = metaRecords.groupMap(_.action)(_.path)
+
+      val deleteRecords = metaRecordsByAction.getOrElse(DELETE, Seq.empty)
+      val addRecords = metaRecordsByAction.getOrElse(ADD, Seq.empty).filter(record => !deleteRecords.contains(record))
+      val otherRecords = metaRecordsByAction.filter(record => record._1 != ADD && record._1 != DELETE).values.flatten
+
+      val notDeletedData = deleteRecords.filter(dataFiles.contains)
+      val missingData = addRecords.diff(dataFiles)
+      val unknownData = dataFiles.diff(addRecords)
+      val isDataIssueDetected =
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/spark_metadata_tool/Application.scala]: `isDataIssueDetected` -> `noDataIssueDetected`

File: src/main/scala/za/co/absa/spark_metadata_tool/MetadataTool.scala [hunk 7]
```diff
@@ -128,6 +129,54 @@ class MetadataTool(io: FileManager) {
...
+   * @param rootDirPath
+   *   path to root directory
+   * @return
+   *   all files in directory and all subdirectories recursively
+   */
+  def listFilesRecursively(rootDirPath: Path): Either[AppError, Seq[Path]] = {
+    for {
+      dirs <- io.listDirectories(rootDirPath)
+      files <- io.listFiles(rootDirPath)
+      filesInSubDir <- dirs.flatTraverse(dir =>  listFilesRecursively(dir))
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/spark_metadata_tool/MetadataTool.scala]: nit: remove one extra space after the arrow

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala (hunk 4 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/spark_metadata_tool/Application.scala:119]: Intended, the same behaviour in every mode.
I didn't introduce it

File: src/main/scala/za/co/absa/spark_metadata_tool/model/MetadataRecord.scala (hunk 1 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/spark_metadata_tool/model/MetadataRecord.scala]: Done

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala (hunk 2 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/spark_metadata_tool/Application.scala]: Done

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala (hunk 3 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/spark_metadata_tool/Application.scala]: Done

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala (hunk 5 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/spark_metadata_tool/Application.scala]: "Const", but Tomas used lowercase so will follow that

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala (hunk 6 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/spark_metadata_tool/Application.scala]: :D Yop fixed

File: src/main/scala/za/co/absa/spark_metadata_tool/MetadataTool.scala (hunk 7 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/spark_metadata_tool/MetadataTool.scala]: fixed

File: src/main/scala/za/co/absa/spark_metadata_tool/Application.scala [hunk 8]
```diff
@@ -121,30 +121,54 @@ object Application extends App {
...
+    }
+    if(unknownData.nonEmpty) {
+      logger.error(s"Unknown data. START")
+      unknownData.foreach(r => logger.error(r.toString))
+      logger.error(s"Unknown data. END")
+    }
+    if(otherRecords.nonEmpty) {
+      logger.error(s"Unknown records in metadata. START")
+      otherRecords.foreach(r => logger.error(r.toString))
+      logger.error(s"Unknown records in metadata. END")
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/spark_metadata_tool/Application.scala:171]: I guess you could have used `.mkString("\n")`, but I don't mind, tbh.

**Review by kevinwallimann — APPROVED** [review]

**Pull request merged**